server/src/main/java/org/elasticsearch/cluster/metadata/ProjectId.java

```diff
@@ -53,8 +53,9 @@ private static boolean isValidIdChar(char c) {
         return (c >= '0' && c <= '9') || (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || c == '_' || c == '-';
     }
 
-    public ProjectId(StreamInput in) throws IOException {
-        this(in.readString());
+    public static ProjectId readFrom(StreamInput in) throws IOException {
+        final var id = in.readString();
+        return Metadata.DEFAULT_PROJECT_ID.id.equals(id) ? Metadata.DEFAULT_PROJECT_ID : new ProjectId(id);
```

Comment:
I am raising this PR as a draft to check whether this is something we want to do. Removal of this constructor will make serverless tests fail. So we need a linked PR if we want to continue on this. If this is desirable, we could push it further to prevent the `new ProjectId(String)` constructor from using "default".  To achieve that, we will have to turn `ProjectId` into a class instead of a record so that the constructor can be private. I am open to suggestions.